File: packages/ubaa_app/lib/src/bridge/read/evaluation.dart
part of '../bridge_backend.dart';

Future<FeatureResult> _loadEvaluationOnRoute(
  BridgeBackend backend, {
  required ConnectionMode route,
}) async {
  try {
    final result = await backend.client.evaluationAllOnRoute(
      route: _toBridgeConnectionMode(route),
    );
    return _mapEvaluationResult(
      result.data,
      _toConnectionMode(result.pinnedRoute),
    );
  } on BridgeError catch (error) {
    throw _mapError(error);
  }
}

Future<FeatureResult> _loadEvaluationFeature(
  BridgeBackend backend,
  FeatureId feature,
  FeatureQuery query,
) async {
  if (feature != FeatureId.evaluation) {
    throw StateError('unexpected feature: $feature');
  }
  switch (query.view) {
    case FeatureQueryView.summary:
    case FeatureQueryView.evaluationPending:
      final result = await backend.client.evaluationAll();
      return _mapEvaluationResult(
        result.data,
        _toConnectionMode(result.route.resolvedRoute),
        pendingOnly: query.view == FeatureQueryView.evaluationPending,
      );
    default:
      throw const BackendException(UbaaErrorCode.invalidInput);
  }
}

FeatureResult _mapEvaluationResult(
  BridgeEvaluationCoursesResponse data,
  ConnectionMode resolvedRoute, {
  bool pendingOnly = false,
}) {
  final normalizedTargetCounts = <String, int>{};
  for (final course in data.courses) {
    final key = _normalizedEvaluationBridgeTargetKey(course.submitTarget);
    if (key != null) {
      normalizedTargetCounts.update(
        key,
        (count) => count + 1,
        ifAbsent: () => 1,
      );
    }
  }
  final courses = pendingOnly
      ? data.courses.where((course) => !course.isEvaluated)
      : data.courses;
  final details = courses
      .map(
        (course) => _mapEvaluationCourseDetail(course, normalizedTargetCounts),
      )
      .toList(growable: false);
  final progress = data.progress;
  final summary = pendingOnly
      ? '待评 ${progress.pendingCourses} 门'
      : '已评 ${progress.evaluatedCourses}/${progress.totalCourses} 门';
  return FeatureResult.success(
    summary: summary,
    details: details,
    resolvedRoute: resolvedRoute,
  );
}

FeatureDetail _mapEvaluationCourseDetail(
  BridgeEvaluationCourse course,
  Map<String, int> normalizedTargetCounts,
) {
  final target = _mapEvaluationSubmitTarget(course.submitTarget);
  final targetKey = target?.selectionKey;
  final originalEligibility = _toEvaluationActionEligibility(
    course.submitEligibility,
  );
  final allowedIsConsistent =
      originalEligibility == ActionEligibility.allowed &&
      target != null &&
      !course.isEvaluated &&
      course.kcmc.trim().isNotEmpty &&
      course.bpmc.trim().isNotEmpty &&
      course.id == _evaluationCourseId(target) &&
      normalizedTargetCounts[targetKey] == 1;
  final eligibility = allowedIsConsistent
      ? ActionEligibility.allowed
      : originalEligibility == ActionEligibility.allowed ||
            course.submitTarget != null
      ? ActionEligibility.unknown
      : originalEligibility;
  final safeTarget = allowedIsConsistent ? target : null;
  final courseName = course.kcmc.trim().isEmpty ? '未知课程' : course.kcmc;
  final teacherName = course.bpmc.trim().isEmpty ? '未知教师' : course.bpmc;
  return FeatureDetail(
    title: courseName,
    subtitle: teacherName,
    fields: _compactFields(<FeatureField?>[
      _field('状态', course.isEvaluated ? '已评' : '待评'),
      course.id.trim().isEmpty ? null : _field('课程 ID', course.id),
    ]),
    actions: <FeatureAction>[
      EvaluationSubmitAction(eligibility: eligibility, target: safeTarget),
    ],
  );
}

EvaluationSubmitTarget? _mapEvaluationSubmitTarget(
  BridgeEvaluationSubmitTarget? target,
) {
  if (target == null ||
      target.rwid.isEmpty ||
      target.rwid.trim() != target.rwid ||
      target.wjid.isEmpty ||
      target.wjid.trim() != target.wjid ||
      target.kcdm.isEmpty ||
      target.kcdm.trim() != target.kcdm ||
      target.bpdm == '' ||
      (target.bpdm != null && target.bpdm!.trim() != target.bpdm)) {
    return null;
  }
  return EvaluationSubmitTarget(
    rwid: target.rwid,
    wjid: target.wjid,
    kcdm: target.kcdm,
    bpdm: target.bpdm,
  );
}

String? _normalizedEvaluationBridgeTargetKey(
  BridgeEvaluationSubmitTarget? target,
) {
  if (target == null) return null;
  final rwid = target.rwid.trim();
  final wjid = target.wjid.trim();
  final kcdm = target.kcdm.trim();
  if (rwid.isEmpty || wjid.isEmpty || kcdm.isEmpty) return null;
  final bpdm = target.bpdm?.trim();
  return EvaluationSubmitTarget(
    rwid: rwid,
    wjid: wjid,
    kcdm: kcdm,
    bpdm: bpdm == null || bpdm.isEmpty ? null : bpdm,
  ).selectionKey;
}

String _evaluationCourseId(EvaluationSubmitTarget target) =>
    '${target.rwid}_${target.wjid}_${target.kcdm}_${target.bpdm ?? ''}';

ActionEligibility _toEvaluationActionEligibility(
  BridgeActionEligibility eligibility,
) => switch (eligibility) {
  BridgeActionEligibility.allowed => ActionEligibility.allowed,
  BridgeActionEligibility.denied => ActionEligibility.denied,
  BridgeActionEligibility.unknown => ActionEligibility.unknown,
};

File: crates/ubaa-flutter-bridge/src/api/read/evaluation.rs
//! 教学评教安全公开投影与跨课程一致性校验。

use std::collections::HashMap;

use ubaa_core::facade as domain;

use super::{
    BridgeActionEligibility, BridgeEvaluationCourse, BridgeEvaluationCoursesResponse,
    BridgeEvaluationProgress, BridgeEvaluationSubmitTarget,
};

pub(super) fn map_evaluation(
    response: domain::EvaluationCoursesResponse,
) -> BridgeEvaluationCoursesResponse {
    let mut target_counts = HashMap::new();
    for target in response
        .courses
        .iter()
        .filter_map(|course| course.submit_target.clone())
        .filter_map(normalize_evaluation_target_identity)
    {
        *target_counts.entry(target).or_insert(0_usize) += 1;
    }
    BridgeEvaluationCoursesResponse {
        courses: response
            .courses
            .into_iter()
            .map(|course| map_evaluation_course(course, &target_counts))
            .collect(),
        progress: BridgeEvaluationProgress {
            total_courses: response.progress.total_courses,
            evaluated_courses: response.progress.evaluated_courses,
            pending_courses: response.progress.pending_courses,
        },
    }
}

fn map_evaluation_course(
    course: domain::EvaluationCourse,
    target_counts: &HashMap<BridgeEvaluationSubmitTarget, usize>,
) -> BridgeEvaluationCourse {
    let mut submit_eligibility = map_action_eligibility(course.submit_eligibility);
    let submit_target = course.submit_target.and_then(map_evaluation_target);
    let allowed_target_is_consistent = submit_target.as_ref().is_some_and(|target| {
        !course.is_evaluated
            && !course.kcmc.trim().is_empty()
            && !course.bpmc.trim().is_empty()
            && course.id == evaluation_course_id(target)
            && target_counts.get(target) == Some(&1)
    });
    let submit_target = match submit_eligibility {
        BridgeActionEligibility::Allowed if allowed_target_is_consistent => submit_target,
        BridgeActionEligibility::Denied | BridgeActionEligibility::Unknown
            if submit_target.is_none() =>
        {
            None
        }
        _ => {
            submit_eligibility = BridgeActionEligibility::Unknown;
            None
        }
    };
    BridgeEvaluationCourse {
        id: course.id,
        kcmc: course.kcmc,
        bpmc: course.bpmc,
        is_evaluated: course.is_evaluated,
        submit_eligibility,
        submit_target,
    }
}

fn evaluation_course_id(target: &BridgeEvaluationSubmitTarget) -> String {
    format!(
        "{}_{}_{}_{}",
        target.rwid,
        target.wjid,
        target.kcdm,
        target.bpdm.as_deref().unwrap_or_default(),
    )
}

fn map_evaluation_target(
    target: domain::EvaluationSubmitTarget,
) -> Option<BridgeEvaluationSubmitTarget> {
    let required = [&target.rwid, &target.wjid, &target.kcdm];
    if required
        .into_iter()
        .any(|value| value.is_empty() || value.trim() != value)
    {
        return None;
    }
    let bpdm = match target.bpdm {
        None => None,
        Some(value) if value.is_empty() => None,
        Some(value) if value.trim() == value => Some(value),
        Some(_) => return None,
    };
    Some(BridgeEvaluationSubmitTarget {
        rwid: target.rwid,
        wjid: target.wjid,
        kcdm: target.kcdm,
        bpdm,
    })
}

fn normalize_evaluation_target_identity(
    target: domain::EvaluationSubmitTarget,
) -> Option<BridgeEvaluationSubmitTarget> {
    let rwid = target.rwid.trim().to_owned();
    let wjid = target.wjid.trim().to_owned();
    let kcdm = target.kcdm.trim().to_owned();
    if rwid.is_empty() || wjid.is_empty() || kcdm.is_empty() {
        return None;
    }
    let bpdm = target
        .bpdm
        .map(|value| value.trim().to_owned())
        .filter(|value| !value.is_empty());
    Some(BridgeEvaluationSubmitTarget {
        rwid,
        wjid,
        kcdm,
        bpdm,
    })
}

fn map_action_eligibility(value: domain::ActionEligibility) -> BridgeActionEligibility {
    match value {
        domain::ActionEligibility::Allowed => BridgeActionEligibility::Allowed,
        domain::ActionEligibility::Denied => BridgeActionEligibility::Denied,
        domain::ActionEligibility::Unknown => BridgeActionEligibility::Unknown,
    }
}
